[WIP] SIF

henriman opened the pull request:
PR for verifying secure information flow for the VerifiedSCION router; see #409 for more information.

---

TODOs (besides what is mentioned in the issue mentioned above)

- [ ] At the moment, this branch still contains the assumption `path.AbsMac == io.AbsMac` in `router/io-spec-atomic-events.gobra`. 
  - The way the project is set up at the moment, one cannot use `path.AbsMac` in `io/io-spec.gobra` since since importing `pkg/slayers/path` would lead to a cyclical import. Thus, at the moment, I define another `AbsMac` function in `io/tmp_defs.gobra` and assume that this performs the same computation.
  - However, we decided that it would likely make more sense to move the relevant definitions in `pkg/slayers/path/io_msgterm_spec.gobra` to `io`.

henriman committed:
Add MWE for proving SIF

henriman committed:
Add MWE2

henriman committed:
Address PR feedback

henriman committed:
Fix package names and remove state.lastMsg3

jcp19 committed:
Merge branch 'master' into master

jcp19 committed:
Merge branch 'master' into master

henriman committed:
Extend io-spec with declassification spec

henriman committed:
Fix declassification spec

henriman committed:
Fix adding decl to dp3s_iospec_ordered

henriman committed:
Start implementing IOD spec

What is missing is the `AtomicDecl` and `InternalDecl` event in
`io-spec-atomic-events.gobra` and `io-spec-abstract-transitions.gobra`.

henriman committed:
Start implementing classification spec

Pre- and postconditions still need to be "lifted" to the atomic and
internal events. `WriteTo` may also need a spec

henriman committed:
Merge remote-tracking branch 'base/master' into sif

henriman committed:
Merge branch 'sif' into sif-io-spec

henriman committed:
Mostly finished implementing security policy

More data needs to be asserted low, of course.
`DeclEvent` needs to be verified once slayers problem is worked out

henriman committed:
Verify private/topology/underlay

henriman committed:
Finish implementing security policy

`DeclEvent` was removed as it just copied `AtomicDecl`. It seems that we
don't need to perform that much translation of the pre- and
postconditions as the other events.

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Add additional annotations to private/topology/underlay

In particular, add postconditions asserting that results are low
wherever it is applicable, in order to make the contracts more precise.
This is in anticipation of needing this information for further
verification.

Note that I still have some concerns regarding the pre- and
postconditions of `fmt.Sprintf` and `serrors.New` which are elaborated
in comments.

henriman committed:
Add minimum annotations to verify private/topology

henriman committed:
Add appropriate postconditions to private/topology

As for private/topology/underlay, I have added these to make the
contracts more precise, as I anticipate I will need this for further
verification

Note that at the moment, we still need a lot of `assume` statements due
to Gobra issues #831 and #832. Also, `LinkType.String` doesn't have an
appropriate postcondition yet, as I couldn't assert (nor assume) low(err.Error()) for
a low `err`.

henriman committed:
Add `low` precondition to `fmt.Sprintf`

I forgot to require all elements of `v` to be low. This also required
the addition of another assumption in `private/topology/underlay`, as
Gobra can't infer that all elements of a slice created from low values
must be low (cf. Gobra issue #835)

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Add assumption due to Gobra issue #835

henriman committed:
Add minimum annotations to pkg/addr

There still remain a lot of assumptions due to Gobra issues

henriman committed:
Remove MWEs here

henriman committed:
Merge branch 'sif' into sif-io-spec

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Clean up and minor fixes and improvements

Most notably sensitivity requirements in interfaces have been made
uniform

henriman committed:
Add additional postconditions to pkg/addr/fmt.go

Note that the way `strings.Builder` is annotated may not be final

henriman committed:
Change sensitivity requirements of mt.Sprintf and serrors.New

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Change pkg/slayers/doc.go to prevent parsing error

doc.go contains a multi-line comment which includes "package
benchmarks". At least under my Windows config, this led to Gobra
incorrectly parsing it as belonging to a package called `benchmarks`
(instead of `slayers`).

henriman committed:
Merge branch 'sif' into sif-io-spec

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Introduce functions to express assumptions

Together with the assumptions, I also moved `LowBytes` to a dedicated
folder: `verification/util/sif`

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

jcp19 commented on pkg/slayers/scion_spec.gobra:
```diff
@@ -412,20 +412,21 @@ requires acc(sl.Bytes(ub, 0, len(ub)), _)
 func (s *SCION) EqAbsHeader(ub []byte) bool {
 	return unfolding acc(s.Mem(ub), _) in
-		let low := CmnHdrLen+s.AddrHdrLenSpecInternal() in
+		// SIF: rename due to conflict with low assertion
```
No need for these comments, this is ghost anyway

jcp19 commented on pkg/slayers/scion_spec.gobra:
```diff
@@ -442,10 +443,11 @@ requires acc(s.Mem(ub), _)
 func (s *SCION) ValidScionInitSpec(ub []byte) bool {
 	return unfolding acc(s.Mem(ub), _)                  in
-		let low := CmnHdrLen+s.AddrHdrLenSpecInternal() in
+		// SIF: rename due to conflict with low assertion
```
```suggestion
```

jcp19 reviewed (commented):
Please merge the master branch here and fix the conflicts

jcp19 commented on router/dataplane.go:
```diff
@@ -145,12 +145,21 @@ type BatchConn interface {
 	// @ ensures   err == nil ==>
 	// @ 	forall i int :: { &msgs[i] } 0 <= i && i < n ==>
 	// @ 		MsgToAbsVal(&msgs[i], ingressID) == old(MultiReadBioIO_val(place, n)[i])
+	// SIF: classification spec
+	// NOTE: The postcondition for recv specifies that received messages are low.
+	// TODO: I have decided to mark the abstraction as low for now.
+	// And also used MultiReadBioIOI_val instead of MsgToAbsVal to match what is
+	// used in permissions.
```
```suggestion
	// Classification spec for SIF.
	// The received messages are low, i.e., observable by the attacker.
	// TODO: we mark the abstraction as low for now and used MultiReadBioIOI_val
	// instead of MsgToAbsVal to match what is used in permissions.
```

jcp19 replied:
it is not clear how one is supposed to address this TODO

jcp19 commented on router/dataplane.go:
```diff
@@ -145,12 +145,21 @@ type BatchConn interface {
 	// @ preserves acc(sl.Bytes(b, 0, len(b)), R10)
 	// @ ensures   err == nil ==> 0 <= n && n <= len(b)
 	// @ ensures   err != nil ==> err.ErrorMem()
+	// SIF: add to classification spec later as well (bfdSend)
```
For these kinds of TODOs, I'd rather open an issue on VerifiedSCION's issue tracker

jcp19 commented on router/dataplane.go:
```diff
@@ -161,6 +170,10 @@ type BatchConn interface {
 	// preconditions for IO-spec:
 	// @ requires  MsgToAbsVal(&msgs[0], egressID) == ioAbsPkts
 	// @ requires  io.token(place) && io.CBioIO_bio3s_send(place, ioAbsPkts)
+	// SIF: classification spec
+	// NOTE: I mark `ioAbsPkts` instead of `MsgToAbsVal(...)` as low here to
+	// match variable used in send permission.
```
```suggestion
	// Classification spec:
	// NOTE: `ioAbsPkts` is marked as low instead of `MsgToAbsVal(...)` to
	// match variable used in send permission.
```

jcp19 committed:
Delete verification/scratch/mwe2/mwe2.gobra

jcp19 committed:
Delete verification/scratch/mwe1/mwe1.gobra

jcp19 commented on verification/io/io-spec.gobra:
```diff
@@ -318,4 +369,14 @@ requires token(t) && CBio_IN_bio3s_exit(t, v)
+ghost
+decreases
+requires token(t) && CBio_IN_bio4s_decl(t, v)
+requires v.isIO_decl_val && low(v.IO_decl_val_inif) && low(v.IO_decl_val_egif) && low(v.IO_decl_val_ts) && low(v.IO_decl_val_beta)
```
please break this big expression in separate lines to make it easier to read

jcp19 commented on verification/io/values.gobra:
```diff
@@ -52,4 +52,17 @@ type IO_val adt {
+		IO_decl_val_ts uint
+		IO_decl_val_inif option[IO_ifs]
+		IO_decl_val_egif option[IO_ifs]
+		IO_decl_val_sigma IO_msgterm
```
this field seems unnecessary

jcp19 replied:
Please add a comment here explaining why we are keeping the field `sigma` even though it is logically redundant

jcp19 replied:
Ping for me to check

jcp19 commented on verification/io/io-spec.gobra:
```diff
@@ -318,4 +369,14 @@ requires token(t) && CBio_IN_bio3s_exit(t, v)
 func Exit(ghost t Place, ghost v IO_val)
 
+// SIF: declassification action
+// TODO: maybe it's a good idea to split up sigma and p in IO_val
```
not clear what you mean by this TODO

henriman committed:
Merge branch 'sif-io-spec' of https://github.com/henriman/VerifiedSCION into sif-io-spec

henriman committed:
Improve style

henriman committed:
Align pre- and postconditions

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into sif-io-spec

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Adjust to sif.LowBytes

henriman committed:
Fix typo

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into sif-private-topology

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into sif-pkg-addr

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into sif-private-underlay-conn

henriman committed:
Verify private/underlay/conn

henriman committed:
Verify pkg/slayers/path/empty except for implementation proof

henriman committed:
Address PR feedback and clean up

henriman committed:
Declassify concrete instead of abstract MAC tag

I couldn't verify my changes to `io-spec-atomic-events.gobra` due to a
magic wand in `dataplane.go`. Also, I couldn't import `pkg/slayers/path`
in `io` as this leads to a cyclical import; thus, for the moment, I
copied the relevant definitions in `io/tmp_defs.gobra` and assume that
they equal the ones in `io_msgterm_spec.gobra`.

henriman committed:
Use `trusted` to suppress problems with closure implementation proof for now

henriman committed:
Verify pkg/slayers/path

At the moment, I'm using pure functions to access fields due to Gobra
issue 883. Maybe this should be changed once that bug is resolved.

henriman committed:
Verify pkg/slayers/path/onehop

henriman committed:
Verify pkg/slayers/path/epic

henriman committed:
Merge branch 'sif-private-topology' of https://github.com/henriman/VerifiedSCION into sif-private-topology

henriman committed:
Fix typo

henriman committed:
Clean up

- Still contains workaround for bug 835/890
- Still contains workaround for missing feature 846

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Fix contract of `serrors.New`

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Clean up

henriman committed:
Rename `AssumeLowSliceToLowString` to `LowSliceImpliesLowString`

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Stop using `sif.LowBytes`

henriman committed:
Clean up

henriman committed:
Use abstract pure Boolean functions instead of `LowMem` etc.

henriman committed:
Clean up

henriman committed:
Fix contract of `TextUnmarshaler.UnmarshalText`

henriman committed:
Merge branch 'sif-pkg-addr' into sif-private-underlay-conn

henriman committed:
Address feedback

henriman committed:
Address missed feedback

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into sif-io-spec

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Minor style adjustments

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Make verification/dependencies/net verify

henriman committed:
Merge branch 'sif-pkg-addr' into sif-pkg-addr-rework

henriman committed:
Minor style adjustments

henriman committed:
Merge branch 'sif-pkg-addr-rework' into sif-private-underlay-conn

henriman committed:
Explain need for longer names for `IsLow`

... in `fmt.Stringer` and `flag.Value`

henriman committed:
Replace Low predicates with abstract IsLow functions

henriman committed:
Adjust annotations of serrors

Previously we required all elements of `errCtx` to be low. Aside from
the fact that this annotation didn't make sense in case `errCtx`
contained values of reference types, it was also unnecessary, since (by
now) we only require that whether the resulting error is not nil to be
low -- which was already guaranteed by the contract.

We do, however, need to ensure that there is no non-low control flow.
This might require further annotations in the future.

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Remove workaround for Gobra issue #835/#890

Due to the change in the contract of serrors, the workaround is not
necessary anymore

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Adjust annotations for serrors

This also allowed the removal of several workarounds for bugs 835/890

henriman committed:
Remove unnecessary annotations

henriman committed:
Merge branch 'sif-pkg-addr' into sif-pkg-addr-rework

henriman committed:
Merge branch 'sif-pkg-addr-rework' into sif-private-underlay-conn

henriman committed:
Fix sensitivity annotations of pointers

henriman committed:
Clean up

henriman committed:
Add some clarifications

henriman committed:
Fix merge

henriman committed:
Clean up

henriman committed:
Merge branch 'sif-pkg-slayers-path-empty' into sif-pkg-slayers-path

henriman committed:
Clean up

henriman committed:
Merge branch 'sif-pkg-slayers-path' into sif-pkg-slayers-path-epic

henriman committed:
Rework sensitivity in interfaces

henriman committed:
Improve stability and make BatchConn match

henriman committed:
"Introduce body" to `isZeros` and improve stability of `NewReadMessages`

`isZeros` now has accompanying `isZeroSpec` with implementation s.t. one
can infer `low(isZeros(s))` if the elements of `s` are low.

`NewReadMessages` has different triggers to hopefully improve stability
(Joao's suggestion). Cannot really test yet, since package is now
failing (cf. Gobra issue #948)

henriman committed:
Avoid requiring input of `fmt.Sprintf` to be low

As `Sprintf` takes input of type `interface{}`, which may also contain
pointer types, it is not clear how to require the input to be low.

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into sif-io-spec

henriman committed:
Merge branch 'sif-io-spec' into sif-private-topology-underlay

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr-rework

henriman committed:
Merge branch 'sif-pkg-addr' into sif-private-underlay-conn

henriman committed:
Merge branch 'sif-private-underlay-conn' into sif-pkg-slayers-path-empty

henriman committed:
Merge branch 'sif-private-underlay-conn' into sif-pkg-slayers-path

henriman committed:
Add comments

henriman committed:
Clean up

henriman committed:
Add //@ before assert

henriman committed:
Clean up

henriman committed:
Clean up

henriman committed:
Merge branch 'sif-pkg-addr' into sif-private-underlay-conn

henriman committed:
Clean up further and add notes (NOTE[henri])

henriman committed:
Address PR feedback

henriman committed:
Merge branch 'sif-pkg-addr' into sif-private-underlay-conn

I adjusted private/underlay/conn/conn.go, `New` accordingly

henriman committed:
Merge branch 'sif-private-underlay-conn' into sif-pkg-slayers-path-empty

henriman committed:
Merge branch 'sif-pkg-slayers-path-empty' into sif-pkg-slayers-path

henriman committed:
Minor adjustments

henriman committed:
Merge branch 'sif-pkg-slayers-path' into sif-pkg-slayers-path-epic

henriman committed:
Clean up

henriman committed:
Make pkg/slayers/path/empty verify again

henriman committed:
Further clean up

henriman committed:
Add explicit contract to `TrimSuffix`

henriman committed:
Adapt pure annotations elsewhere

henriman committed:
Adapt pure annotations that I missed

henriman committed:
Address PR feedback and add `&& low(i) ==>`

henriman committed:
Merge branch 'sif-private-topology-underlay' into sif-private-topology

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Make contract of `net.IP.To4` less implementation-specific

henriman committed:
Add `&& low(i) ==>` in quantified assertions where appropriate

henriman committed:
Clean up

henriman committed:
Remove `&& low(i) ==>` from one assertion not dealing with sensitivity

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Remove occurrences of `&& low(i) ==>` added accidentally

henriman committed:
Merge branch 'sif-pkg-addr' into sif-private-underlay-conn

henriman committed:
Address PR feedback

henriman committed:
Add `&& low(i) ==>`

henriman committed:
Refer to Gobra issue #957 on NewReadMessages

henriman committed:
Remove assumption for Gobra issue 831

Issue 831 is fixed by PR 961

henriman committed:
Merge branch 'sif-private-topology' into sif-pkg-addr

henriman committed:
Add TODO to IsLow regarding adding hyper

henriman committed:
Add TODO to IsLow regarding adding hyper

henriman committed:
Merge branch 'sif-pkg-addr' into sif-private-underlay-conn

henriman committed:
Add TODO regarding hyper to DataPlane as well

henriman committed:
Merge branch 'sif-private-underlay-conn' into sif-pkg-slayers-path-empty

henriman committed:
Merge branch 'sif-pkg-slayers-path-empty' into sif-pkg-slayers-path

henriman committed:
Address PR feedback

henriman committed:
Merge branch 'sif-pkg-slayers-path' into sif-pkg-slayers-path-epic

henriman committed:
Add TODO to IsLow regarding adding hyper

henriman committed:
Merge branch 'sif-pkg-slayers-path-empty' into sif-pkg-slayers-path-onehop

henriman committed:
Use abstract functions instead of predicate and clean up

henriman committed:
Merge pull request #1 from henriman/sif-private-topology-underlay

Verify `private/topology/underlay`

henriman committed:
Align contract clauses

henriman committed:
Merge pull request #2 from henriman/sif-private-topology

Verify `private/topology`

henriman committed:
Remove addressed NOTE[henri] comments

henriman committed:
Merge pull request #5 from henriman/sif-pkg-addr

Verify `pkg/addr`

henriman committed:
Merge branch 'sif-io-spec' into sif-private-underlay-conn

henriman committed:
Address PR feedback

henriman committed:
Remove `trusted` from `UDPConn.IsLow`

As `UDPConn` is "abstract" (i.e. only defined using `PrivateField`),
the issue here is not of not being able to implement `IsLow`, only that
we still need to add `hyper` to it. Update comment accordingly.

henriman committed:
Fix comment on `NewReadMessages`

henriman committed:
Merge pull request #6 from henriman/sif-private-underlay-conn

Verify `private/underlay/conn`

henriman committed:
Merge branch 'sif-io-spec' into sif-pkg-slayers-path-empty

henriman committed:
Merge pull request #7 from henriman/sif-pkg-slayers-path-empty

Verify `pkg/slayers/path/empty`

henriman committed:
Merge branch 'sif-io-spec' into sif-pkg-slayers-path

henriman committed:
Merge pull request #8 from henriman/sif-pkg-slayers-path

Verify `pkg/slayers/path`

henriman committed:
Merge branch 'sif-io-spec' into sif-pkg-slayers-path-epic

henriman committed:
Merge pull request #9 from henriman/sif-pkg-slayers-path-epic

Verify `pkg/slayers/path/epic`

henriman committed:
Merge branch 'sif-io-spec' into sif-pkg-slayers-path-onehop

henriman committed:
Clean up

henriman committed:
Merge pull request #12 from henriman/sif-pkg-slayers-path-onehop

Verify `pkg/slayers/path/onehop`
